### PR TITLE
Avoid leaking transports cache

### PR DESCRIPTION
**Pull request by aruiz14**
## Problem

**TL;DR** client-go 1.26 changed the semantics of `kubernetes.NewForConfig(restConfig)` and is now adding transports to an internal cache without a pruning mechanism: Rancher leaks memory. [A client-go patch to optionally restore the old behavior was rejected](https://github.com/kubernetes/kubernetes/issues/125818#issuecomment-2314686148), this PR contains a workaround in Rancher code.

**Details**: client-go 1.26+ changed the semantics of `kubernetes.NewForConfig(restConfig)` by caching TLS transport objects in some cases which were not cached in earlier versions (notably, configurations with non-nil `Dial` functions, see https://github.com/kubernetes/kubernetes/issues/125818). 

As client-go never prunes its internal TLS transport cache, we have observed this cache to grow indefinitely in some customer environments, in particular, using downstream cluster in AKS (see https://github.com/rancher/rancher/issues/45509).

Specifically, when `rest.Config`'s `Dial` field is set, `rest.Config` is transformed to a `transport.Config` via [`TransportConfig()`](https://pkg.go.dev/k8s.io/client-go@v0.31.0/rest#Config.TransportConfig), which wraps the `Dial` function into a `transport.DialHolder` struct. Golang functions are not comparable, so they had to wrap it in a struct and take its pointer instead, which is indeed comparable. However, since [a new struct is being created on every call to `TransportConfig()`](https://github.com/kubernetes/client-go/blob/v0.31.0/rest/transport.go#L115), this causes the calculated cache key to be different for any consecutive calls to the same `restConfig.TransportConfig()`, ultimately causing cache misses and growing indefinitely (also not LRU-based nor fixed max size).
 
## Solution

Places constructing a `rest.Config` and a Kubernetes client based on it seem to be:
 - [`clustermanager.ToRESTConfig`](https://github.com/rancher/rancher/blob/f1382133ebd844fc1dc63d5425a154f67acad8d4/pkg/clustermanager/manager.go#L315): does not set the `Dial` field under any case, so this does not trigger the problem.
 - Multiple `clustermanager` use `dialer.Factory` to obtain a `ClusterDialer`, then use it on the `rest.Config` used to create a Kubernetes client. This is likely the reason why we only observed this behavior in certain cases.

I've added a new method to the `dialerFactory`, allowing to keep a single `transport.DialHolder` for every cluster, and slightly change the way we construct the Kubernetes client so we can directly set this field.

### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: pkg/controllers/management/clusteroperator/utils.go [hunk 1]
```diff
@@ -166,13 +167,39 @@ func (e *OperatorController) CheckCrdReady(cluster *mgmtv3.Cluster, clusterType
...
+		opt(transportConfig)
+	}
+
+	rt, err := transport.New(transportConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	var httpClient *http.Client
+	if rt != http.DefaultTransport || config.Timeout > 0 {
```
Comment (jiaqiluo) [inline pkg/controllers/management/clusteroperator/utils.go:193]: I failed to find the place where the [transport.New(transportConfig)](https://github.com/kubernetes/client-go/blob/v0.30.1/transport/transport.go#L36) function returns an `http.DefaultTransport`, could you provide more context here?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: pkg/controllers/management/clusteroperator/utils.go (hunk 1 above)
Comment (aruiz14) [reply pkg/controllers/management/clusteroperator/utils.go:193]: Yeah, it's quite confusing.
 1. [`kubernetes.NewForConfig(restConfig)`](https://github.com/kubernetes/client-go/blob/v0.30.1/kubernetes/clientset.go#L468), which was used here, ensures the user agent is set, then calls:
 2. [`rest.HTTPClientFor(restConfig)`](https://github.com/kubernetes/client-go/blob/v0.30.1/rest/transport.go#L32), from which I copied these bits in order to be able to access `transportConfig`.

**Review by jakefhyde — DISMISSED** [review]

File: pkg/controllers/management/clusteroperator/utils.go (hunk 1 above)
Comment (jiaqiluo) [reply pkg/controllers/management/clusteroperator/utils.go:193]: Thanks for the links.

**Review by jakefhyde — APPROVED** [review]

**Review by jiaqiluo — DISMISSED** [review]

**Review by moio — DISMISSED** [review]
Left a couple of questions inline, rest LGTM

Thanks!

File: pkg/dialer/factory.go [hunk 2]
```diff
@@ -56,6 +62,27 @@ func (f *Factory) ClusterDialer(clusterName string, retryOnError bool) (dialer.D
...
 
+func (f *Factory) ClusterDialHolder(clusterName string, retryOnError bool) (*transport.DialHolder, error) {
+	f.dialHoldersLock.RLock()
+	cached, ok := f.dialHolders[clusterName]
+	f.dialHoldersLock.RUnlock()
+	if ok {
+		return cached, nil
+	}
+
+	clusterDialer, err := f.ClusterDialer(clusterName, retryOnError)
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply pkg/dialer/factory.go:84]: You are right, nice catch. I've updated it to prevent that specific case, thanks!

> And possibly garbage-collected after use

That wouldn't happen, since the problem is that client-go only ever adds to their transport caches, so it would never be freed.

**Review by moio — APPROVED** [review]
